### PR TITLE
Add OAuth2 Outbound Authenticator Changes 

### PR DESCRIPTION
**Pull request by Shalindri**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml [hunk 1]
```diff
@@ -96,5 +96,4 @@
             </plugin>
         </plugins>
     </build>
-</project>
-
+</project>
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml]:
```suggestion
</project>

```

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml [hunk 2]
```diff
@@ -97,4 +97,3 @@
         </plugins>
     </build>
 </project>
-
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/pom.xml:100]: Is this change intentional? Shall we revert this change please?

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 3]
```diff
@@ -61,55 +59,100 @@
...
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.OAUTH2_TOKEN_URL_DP;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.OAUTH2_USER_INFO_URL;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.OAUTH2_USER_INFO_URL_DESC;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.OAUTH2_USER_INFO_URL_DP;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.SCOPE;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.SCOPE_DESC;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.SCOPE_DP;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.TOKEN_TYPE;
+import static org.wso2.carbon.identity.application.authenticator.oauth2.Oauth2GenericAuthenticatorConstants.VAR_TYPE_BOOLEAN;
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Can we please get rid of the static imports?
https://docs.oracle.com/javase/8/docs/technotes/guides/language/static-import.html

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 4]
```diff
@@ -61,55 +59,100 @@
...
-            logger.error("Exception while sending to the login page.", e);
-            throw new AuthenticationFailedException(e.getMessage(), e);
+            String message = "Exception while redirecting to the login page.";
+            logger.error(message, e);
+            throw new AuthenticationFailedException(message, e);
         } catch (OAuthSystemException e) {
-            logger.error("Exception while building authorization code request.", e);
+            String message = "Exception while building authorization code request.";
+            logger.error(message, e);
+            throw new AuthenticationFailedException(e.getMessage(), e);
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: We don't need to log and throw the same message

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 5]
```diff
@@ -119,87 +162,82 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
-            String clientSecret = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CLIENT_SECRET);
-            String redirectUri = authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.CALLBACK_URL);
+            Map<String, String> authenticatorProperties = getAuthenticatorProperties(context);
+            String clientId = getClientId(authenticatorProperties);
+            String clientSecret = getClientSecret(authenticatorProperties);
+            String redirectUri = getCallbackURL(authenticatorProperties, request.getServerName(),
+                    request.getServerPort());
             Boolean basicAuthEnabled = Boolean.parseBoolean(
-                    authenticatorProperties.get(Oauth2GenericAuthenticatorConstants.IS_BASIC_AUTH_ENABLED));
+                    StringUtils.lowerCase(authenticatorProperties.get(IS_BASIC_AUTH_ENABLED)));
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: To parseBoolean why do we need to make the string lowercase?

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 6]
```diff
@@ -119,87 +162,82 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+            throw new AuthenticationFailedException("Failed to process authentication response.", e);
+        } catch (MisconfigurationException e) {
+            logger.error(e.getMessage(), e);
             throw new AuthenticationFailedException(e.getMessage(), e);
         }
-
     }
 
     protected void buildClaims(AuthenticationContext context, String userInfoString)
-            throws ApplicationAuthenticatorException, AuthenticationFailedException {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:151]: To preserve the backward compatibility lets not modify the method signature of public/protected methods

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 7]
```diff
@@ -490,64 +559,91 @@ public String getName() {
...
+                callbackURL = new URL(DEFAULT_PROTOCOL_IDENTIFIER, serverName, serverPort, CALLBACK_URL_DEFAULT)
+                        .toString();
+            } catch (MalformedURLException e) {
+                throw new MisconfigurationException(e.getMessage(), e);
+            }
+            return callbackURL;
+        }
+        return callbackURL;
+    }
+}
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java]: Add new line

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticatorConstants.java [hunk 8]
```diff
@@ -40,41 +39,44 @@
...
-    public static final String AUTH_HEADER_NAME = "Authorization";
     public static final String TOKEN_TYPE = "Bearer ";
-    public static final String HTTP_GET_METHOD = "GET";
-
-    public static final String VAR_TYPE_BOOLEAN = "boolean";
-
-}
+    public static final String DEFAULT_PROTOCOL_IDENTIFIER = "https";
 
+}
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticatorConstants.java]: new line

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java [hunk 9]
```diff
@@ -211,91 +199,76 @@ protected String getToken(String tokenEndPoint, String clientId, String clientSe
...
             }
-        } finally {
-            IdentityIOStreamUtils.closeReader(bufferReader);
+            throw new ApplicationAuthenticatorException(e.getMessage());
         }
-        return stringBuilder.toString();
     }
 
     protected OAuthClientRequest buildTokenRequest(String tokenEndPoint, String clientId, String clientSecret,
-                                                  String state, String code, String redirectUri,
```
Comment (GDRDABARERA) [inline components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:273]: Does this method extended from the AbstractAuthenticator class? Why we removed state parameter?

**Review by GDRDABARERA — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java (hunk 9 above)
Comment (Shalindri) [reply components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:273]: No it's not.

File: components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java (hunk 6 above)
Comment (Shalindri) [reply components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java:151]: This class has extended in two connectors(Naver and Kakao) and this change has affected to Naver connector. 
Since this change is addressing few issues we had with this generic connector, we decided to change that connector as well [PR for the Naver connector change](https://github.com/wso2-extensions/identity-outbound-auth-naver/pull/6 )

**Review by GDRDABARERA — APPROVED** [review]
